gapis/api/vulkan/frame_loop.go

```diff
@@ -196,6 +233,62 @@ func (f *frameLoop) Flush(ctx context.Context, out transform.Writer)    {}
 func (f *frameLoop) PreLoop(ctx context.Context, out transform.Writer)  {}
 func (f *frameLoop) PostLoop(ctx context.Context, out transform.Writer) {}
 
+func (f *frameLoop) rewriteQueueSubmit(ctx context.Context, sb *stateBuilder, cmd *VkQueueSubmit) *VkQueueSubmit {
+	s := sb.newState
+
+	cmd.Extras().Observations().ApplyReads(s.Memory.ApplicationPool())
+	submitCount := cmd.SubmitCount()
+	submitInfos := cmd.pSubmits.Slice(0, uint64(submitCount), s.MemoryLayout).MustRead(ctx, cmd, s, nil)
+
+	for i := uint32(0); i < submitCount; i++ {
+		si := submitInfos[i]
+		cmdBuffers := si.PCommandBuffers().Slice(0, uint64(si.CommandBufferCount()), s.MemoryLayout).MustRead(ctx, cmd, s, nil)
+		cmdCount := si.CommandBufferCount()
+
+		newCmdBuffers := make([]VkCommandBuffer, cmdCount)
+		for j := uint32(0); j < cmdCount; j++ {
+			newCmdBuffers[j] = cmdBuffers[j]
+			// re-record the command buffer if it's not beginned in this frame.
+			if _, ok := f.commandBeginned[cmdBuffers[j]]; !ok {
+				commandbuffer := f.recreateCommandBuffer(ctx, sb, cmdBuffers[j])
```

Comment:
I dont think we want to re-record the command-buffer just before the quuee submit, we should be re-recording it when we are resetting all of the rest of the resources.

Reply:
So the the re-recording only happened:
-  at the resetting step;
-  and only when the resource used in the command buffer has changed;
- and it was not re-recorded in the original application.

Now the questions is how to check if resource in a command buffer has been change.

Reply:
I think I answered this offline but I would stick with 
 - At the resetting step
 - Only if it was submitted during the loop
 - Only if it was not re-recorded before the first submission in the loop.

Reply:
The updated code implements this logic. And also allocate/free the freed/allocated commandbuffer during loop.